patches/tModLoader/Terraria/ModLoader/ContentCache.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Terraria.ModLoader;
+
+/// <summary>
+/// Container object for <see cref="ILoadable"/> instances added for a <see cref="Mod"/>
+/// </summary>
+internal class ContentCache
+{
+	private static event Action OnUnload;
+	private static Dictionary<nint, IList> _cachedContentForAllMods = new();
+
+	internal static bool hasLoadingStarted;
+	internal static bool loadingContent;
+
+	internal static void Unload() {
+		Interlocked.Exchange(ref OnUnload, null)?.Invoke();
+	}
+
+	private static void UnloadStaticCache() {
+		_cachedContentForAllMods.Clear();
+	}
+
+	public static IEnumerable<T> GetContentForAllMods<T>() where T : ILoadable {
+		if (!hasLoadingStarted || loadingContent) {
```

Comment:
Do we need 2 flags? Can we just have `contentLoadingFinished`, which we can clear in `Unload`?